tests/hats_tap/test_tap_schema_db_threading.py

```diff
@@ -0,0 +1,138 @@
+"""Tests for thread-safety of TAPSchemaDatabase."""
+
+import sqlite3
+import threading
+
+import pytest
+from hats_tap.tap_schema_db import TAPSchemaDatabase
+
+
+class TestThreadSafety:
+    """Test that TAPSchemaDatabase is thread-safe."""
+
+    @pytest.fixture
+    def temp_db(self, tmp_path):
+        """Create a temporary database for testing."""
+        db_path = tmp_path / "test_tap_schema.db"
+        db = TAPSchemaDatabase(str(db_path))
+        db.initialize_schema()
+        # Insert some test data
+        db.insert_schema("test_schema", "Test schema")
+        db.insert_table("test_schema", "test_table", "table", "Test table")
+        return db
+
+    def test_multiple_threads_can_query(self, temp_db):
+        """Test that multiple threads can query the database concurrently."""
+        results = []
+        errors = []
+
+        def query_in_thread():
+            """Query the database in a thread."""
+            try:
+                schemas = temp_db.query("SELECT schema_name FROM schemas")
+                results.append(schemas)
+            except Exception as e:
+                errors.append(e)
+
+        # Create multiple threads
+        threads = [threading.Thread(target=query_in_thread) for _ in range(10)]
+
+        # Start all threads
+        for thread in threads:
+            thread.start()
+
+        # Wait for all threads to complete
+        for thread in threads:
+            thread.join()
+
+        # Check that all threads succeeded
+        assert len(errors) == 0, f"Expected no errors, but got: {errors}"
+        assert len(results) == 10, f"Expected 10 results, but got {len(results)}"
+
+        # Check that each result contains the test schema
+        for result in results:
+            assert len(result) == 1
+            assert result[0]["schema_name"] == "test_schema"
+
+    def test_each_thread_has_own_connection(self, temp_db):
+        """Test that each thread has its own connection."""
+        connection_ids = []
+        lock = threading.Lock()
+
+        def get_connection_id():
+            """Get the connection ID in a thread."""
+            temp_db.connect()
+            conn_id = id(temp_db.connection)
+            with lock:
+                connection_ids.append(conn_id)
```

Comment:
```suggestion
            with temp_db:
                temp_db.connect()
                conn_id = id(temp_db.connection)
                with lock:
                    connection_ids.append(conn_id)
```
I think this should clean up our connections once we're done with them

Reply:
That is true. As it is, the finalizers will tidy up the dangling connections at the end of the test suite, and part of the purpose of the test is to make sure that the connections are open simultaneously; I didn't want to close them too early. But your suggestion shouldn't change that, I don't think, because of the lock, and it will make sure they don't stay open past the test.

Reply:
As it turns out, I had to reverse this suggestion, since the test *does* break if each thread closes its connection. The closing needs to wait for the finalizer, which will happen in `.join`.